pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -98,88 +103,217 @@ func (r *MachineSetSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
-func (r *MachineSetSyncReconciler) reconcileMAPIMachineSettoCAPIMachineSet(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet, capiMachineSet *capiv1beta1.MachineSet) (ctrl.Result, error) {
+// reconcileCAPIMachineSetToMAPIMachineSet reconciles a CAPI MachineSet to a
+// MAPI MachineSet.
+//
+// TODO: Platform specific implementation (currently this works only for AWS,
+// we want a switch on platform somewhere).
+// TODO: Put Gets() for Cluster + Template in helper func.
+//
+//nolint:funlen
+func (r *MachineSetSyncReconciler) reconcileCAPIMachineSetToMAPIMachineSet(ctx context.Context, capiMachineSet *capiv1beta1.MachineSet, mapiMachineSet *machinev1beta1.MachineSet) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("function", "reconcileCAPIMachineSetToMAPIMachineSet")
+
+	infraCluster := &awscapiv1beta1.AWSCluster{}
+	clusterNamespacedName := client.ObjectKey{
+		Namespace: capiMachineSet.GetNamespace(),
+		Name:      capiMachineSet.Spec.ClusterName,
+	}
+
+	if err := r.Get(ctx, clusterNamespacedName, infraCluster); err != nil {
+		logger.Error(err, "Failed to get CAPI InfraCluster", "cluster", clusterNamespacedName.Name)
+
+		r.updateSynchronizedConditionWithPatch(
+			ctx, mapiMachineSet, corev1.ConditionFalse,
+			"FailedToGetCAPIInfraCluster", err.Error(),
+		)
+
+		return ctrl.Result{}, fmt.Errorf("failed to get CAPI InfraCluster: %w", err)
+	}
+
+	infraMachineTemplate := &awscapiv1beta1.AWSMachineTemplate{}
+	templateNamespacedName := client.ObjectKey{
+		Namespace: capiMachineSet.Spec.Template.Spec.InfrastructureRef.Namespace,
+		Name:      capiMachineSet.Spec.Template.Spec.InfrastructureRef.Name,
+	}
+
+	if err := r.Get(ctx, templateNamespacedName, infraMachineTemplate); err != nil {
+		logger.Error(err, "Failed to get InfraMachineTemplate", "name", templateNamespacedName.Name)
+
+		r.updateSynchronizedConditionWithPatch(
+			ctx, mapiMachineSet, corev1.ConditionFalse,
+			"FailedToGetCAPIInfraMachineTemplate", err.Error(),
+		)
+
+		return ctrl.Result{}, fmt.Errorf("failed to get InfraMachineTemplate: %w", err)
+	}
+
+	// Convert the CAPI MachineSet and AWS resources to a MAPI MachineSet
+	newMapiMachineSet, warns, err := capi2mapi.FromMachineSetAndAWSMachineTemplateAndAWSCluster(
+		capiMachineSet, infraMachineTemplate, infraCluster,
+	).ToMachineSet()
+
+	if err != nil {
+		r.updateSynchronizedConditionWithPatch(
+			ctx, mapiMachineSet, corev1.ConditionFalse,
+			"FailedToConvertCAPIMachineSetToMAPI", err.Error(),
+		)
+
+		return ctrl.Result{}, fmt.Errorf("failed to convert CAPI MachineSet to MAPI machine set: %w", err)
+	}
+
+	for _, warning := range warns {
+		logger.Info("Warning during conversion", "warning", warning)
+		r.Recorder.Event(mapiMachineSet, corev1.EventTypeWarning, "ConversionWarning", warning)
+	}
+
+	newMapiMachineSet.Spec.Template.Labels = util.MergeMaps(mapiMachineSet.Spec.Template.Labels, newMapiMachineSet.Spec.Template.Labels)
```

Comment:
This also seems to be redundant since we don't persist the newMapiMachineSet, only the spec from it

Reply:
Ugh - I changed this today and forgot to come back to it. 

The API Server rejects the output from the conversion library, which is why we don't just pass `newMapiMachineSet`. 

I've got a TODO in my notes to dig into this - as it's weird. The error is a 'not found' - despite the test env creating & observing updates (e.g. to `status.AuthoritativeAPI`)

Reply:
> This also seems to be redundant since we don't persist the newMapiMachineSet, only the spec from it

I think this is included? We are updating `.Spec.Template.Labels` and persisting everything under `.Spec`?

Reply:
<img width="1392" alt="Screenshot 2024-10-30 at 15 30 45" src="https://github.com/user-attachments/assets/470075b0-50b1-4683-89d5-20179b2b6b84">

It was the namespace having come from the CAPI MachineSet giving a not found, which makes sense 😂 copying over the resource version was also required, but now means we can apply the output of the conversion library! :D

Reply:
Hmm, I would expect the conversion library to set the appropriate namespaces on the way out, perhaps that's a bug?

Reply:
Yeah, it feels like it although I'm not sure how we could currently: 

```go
newMapiMachineSet, warns, err := capi2mapi.FromMachineSetAndAWSMachineTemplateAndAWSCluster(
		capiMachineSet, infraMachineTemplate, infraCluster).ToMachineSet()
```

We don't pass the existing MAPI MachineSet, so unless we passed the namespace (which is totally an option...) then how would we know what to set it to?

Reply:
I've opened https://github.com/openshift/cluster-capi-operator/issues/229 as I don't think we should try fixing the conversion namespace bug in this PR

Reply:
Hmm yes, fair point, will need to look into this, as there are definitely areas within the conversion library where it does force the namespace. Seems like that's being done inconsistently right now

Reply:
yeah, also this controller allows setting of custom namespaces (e.g for testing), if we want to keep that option then that'll inform how we modify the conversion library